Preselection in tda

chellings opened the pull request:
A useful branch by @nathlacroix. I copied that description from a pull request of an older version of this branch in to the project branch, hoping that this text is still mostly up-to-date.

This pull request allows to perform time domain analysis using single shot data.
The analysis is notified of single shot data by passing data_type="singleshot".

- Adds required analysis metadata in mqm.measure_n_qubit_rabi(), ramsey, t1, t2_echo, q_scale, and QuantumExperiment
- Modifies the multiqubittimedomainanalysis to handle singleshot data
- Allows to preselect jointly on several qubits by providing "preselection_qbs" as a dict where the key is the name of the target qubit and the value is a list of qubit which must be in the ground state for that particular qubit (see doc string of _get_preselection_masks() for more information.

Inviting @antsr to review, FYI @stephlazar @nathlacroix

nathlacroix committed:
Move prep_params and datafilter to extract_data

- make prep_params a class attribute
- retrieve prep_params and data_filter in extract_data

nathlacroix committed:
Remove prep_params from get_indices call in multiqubittimedomainanalysis

The cal point indices in cal_points_dict are used in MQTDA for plots only on data for which any preparation readout (e.g. active reset or preselection) has already been removed. therefore the indices should only consider "high level" data

nathlacroix committed:
Add processing of single_shot readout in MultiQubitTimeDomainAnalysis

nathlacroix committed:
Ensure appropriate arguments are passed in metadata for measure_rabi with ssro

- determine data_type
- decide whether data should be classified (if the classified detector is used, no need to classify the single shots)

nathlacroix committed:
Merge branch 'qudev_master' into enh/preselection_for_tda

nathlacroix committed:
Merge branch 'qudev_master' into enh/preselection_for_tda

Commit:
Merge branch 'qudev_master' into enh/preselection_for_tda

# Conflicts:
#	pycqed/analysis_v2/timedomain_analysis.py

Commit:
Merge branch 'qudev_master' into enh/preselection_for_tda

Commit:
Adds parameters needed for doing preselection to the measure n-qubit time domain functions call, and fixes some docstring words in MultiQubit_TimeDomain_Analysis.

nathlacroix committed:
Add possibility to do analyze 2D single shot measurements in TDA

- adapt _get_single_shots_per_qb() to return all shots in 2D array, where n_soft_sweep points is the innermost loop (allows to avoid many "if 2D/ if 1D" when processing the shots)
- adapt process_single_shots() to do preselection and proper reshaping when dealing with 2D single shot data

nathlacroix committed:
Change parameter "classify" to "classified_ro"

Ensures we have one consistent way of designating that the readout uses a classifier in single qubit and two qubit gate calibrations. Note that classified_ro = logical_inverse(classify).

nathlacroix committed:
Enforce cal_state_rotations for both qubits in Cphase calib

nathlacroix committed:
Change terminology from "classify" to "predict_proba" when processing single shots

- predict_proba= whether or not to apply the classifier on the single shots to obtain probabilities. If false, then it will just average them (if there is preselection then it will just predict the probas on the preselection readouts). If true, then it transforms the raw voltages to probabilies and then averages.
- default to False

nathlacroix committed:
Enable global pre-selection

- add function _get_preselection_masks() which computes the masks based on "preselection_qbs", which is any combination of qubits given by user (defaults to preselecting qubits independently)
- create label with percentage of data kept after preselection
- move gate list parsing in extract_data()

nathlacroix committed:
Add preselection percentage to plot

nathlacroix committed:
Complete doc string

nathlacroix committed:
tda: Add single_shots_per_qb and new name single_shots_per_qb_probs in proc_data_dict

chellings committed:
bugfix MultiCZgate_Calib_Analysis: fixed printing of preselection
- avoid printing it twice (once on top of the dynphase)
- avoid a crash due to parsing error related to \\pm

chellings committed:
Merge branch 'qudev_master' into enh/preselection_for_tda

chellings committed:
MultiQubit_TimeDomain_Analysis: fixed a merge bug

chellings committed:
Merge branch 'qudev_master' into enh/preselection_for_tda

chellings committed:
Merge branch 'qudev_master' into enh/preselection_for_tda

chellings committed:
Revert "Enforce cal_state_rotations for both qubits in Cphase calib" for the following reason.

The user might want to do two-state rotation to get a better estimate of the conditional phase by excluding any f-level preparation or readout errors.
A follow-up commit will allow the user to overwrite the default cal_states_rotations via a keyword argument.

This reverts commit 36ba17ec4c40c959ec7f4f7e419f1ee2d27eddd2.

chellings committed:
CPhase: allow the user to overwrite the default cal_states_rotations via a keyword argument

chellings commented:
@nathlacroix I found unresolved comments in an old pull request that merged an older version of this branch into the project branch. Please check these comments and address them if necessary before this PR gets merged into the master. Thank you! https://github.com/QudevETH/PycQED_py3/pull/143

@nathlacroix As a follow-up on our discussion on cal_states_rotations and after getting some input from @stephlazar, I added the commits 9323adb and d075b18 

@nathlacroix @antsr It would be good to merge this PR soon since it blocks the creation of some other PRs (due to potential merge conflicts)

stephlazar commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -435,11 +437,25 @@ def process_data(self):
             # for now assuming the same for all qubits.
             self.cal_states_dict = self.cp.get_indices(
-                self.qb_names, prep_params)[self.qb_names[0]]
+                self.qb_names)[self.qb_names[0]]
```
why were the prep_params removed here? Will the indices be extracted correctly if preselection or active reset?

nathlacroix replied:
Explanation is in git commit message:
Remove prep_params from get_indices call in multiqubittimedomainanalysis  
The cal point indices in cal_points_dict are used in MQTDA for plots only on data for which any preparation readout (e.g. active reset or preselection) has already been removed. therefore the indices should only consider "high level" data

stephlazar replied:
I'm not sure I understand this comment. There is no cal_points_dict variable/attribute. Did you mean cal_states_dict? If so, that attribute is used to determine how to rotate the data (indices of the cal states). So it sounds as though it needs to take into account whether some data indices correspond to preselection or active reset, as in the image below.

![image](https://user-images.githubusercontent.com/14844492/133994963-ac0c890f-a859-4658-91aa-85902fb42110.png)

stephlazar commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -1106,6 +1125,283 @@ def get_latex_prob_label(prob_label):
+                # make 2D array in case only one channel (1D array)
+                elif np.ndim(shots_per_qb[qbn]) == 1:
+                    shots_per_qb[qbn] = np.expand_dims(shots_per_qb[qbn],
+                                                       axis=-1)
```
There is an indentation error inside the for loop. Is there an if-statement missing or can we just unindent it?

nathlacroix replied:
solved by now :+1:

stephlazar commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -6645,6 +6952,21 @@ def process_data(self):
+                    default_preselection_qbs[qbn].extend(gate_qbs)
+        preselection_qbs = self.get_param_value("preselection_qbs",
+                                                default_preselection_qbs)
+        self.options_dict.update({"preselection_qbs": preselection_qbs})
```
Why is the above extraction not done in the base MultiCZGate_Calib_Analysis? This looks like it should also be added to DynamicPhaseAnalysis.

Commit:
Merge branch 'qudev_master' into enh/preselection_for_tda

stephlazar reviewed (commented):
@chellings you said you fixed something in the DynamicPhaseAnalysis but I didn't see any changes in that class in this PR.

Commit:
MultiCZgate_Calib_Analysis.extract_data: removes unnecessary lines of code

antsr commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -412,16 +415,15 @@ def create_meas_results_per_qb(self):
+            elif "preselection" in self.prep_params.get('preparation_type',
+                                                        'wait'):
                 self.data_filter = lambda x: x[1::2]  # filter preselection RO
         if self.data_filter is None:
             self.data_filter = lambda x: x
```
Should be moved to the `else` case inside previous if statement

chellings replied:
I do not think so since self.data_filter can be modified inside the previous if statement.

antsr commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -1106,6 +1125,283 @@ def get_latex_prob_label(prob_label):
+        Gets single shots from the proc_data_dict and arranges
+        them as arrays per qubit
+        Args:
+            raw (bool): whether or not to return  raw shots (before
```
double space

antsr commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -1106,6 +1125,283 @@ def get_latex_prob_label(prob_label):
+                    assert self.get_param_value("TwoD", False) == True, \
+                        "'TwoD' is False but single shot data seems to be 2D"
+                    n_vn = shots_per_qb[qbn].shape[-1]
+                    n_vn = shots_per_qb[qbn].shape[-1]
```
duplicate row

antsr commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -1106,6 +1125,283 @@ def get_latex_prob_label(prob_label):
+
+        return shots_per_qb
+
+    def _get_preselection_masks(self, presel_shots_per_qb, preselection_qbs=None,
+                                predict_proba=True,
+                                classifier_params=None,
+                                preselection_state_int=0):
```
How does this play together with the `data_filter` lambda function? Is it applied before or after? Should it maybe be merged into the data filter? At least a comment in the docstring about this would be nice.

nathlacroix replied:
good question: for now  this function "replaces" the use of the data_filter. There are several short comings to the data_filter: does not store the mask of which data is kept or not, and also does not work on individual qubits but is the same for all qubits. I do agree that in the future, it would be better to improve the data_filter capabilities and perhaps merge these together, but this would require significant effort. I added a note in the doc string though.

antsr commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -1106,6 +1125,283 @@ def get_latex_prob_label(prob_label):
+        average the shots/probabilities.
+
+        Args:
+            predict_proba (bool): whether or not to assign probabilities to shots.
+                If True, it assumes that shots in the proc_data_dict are the
+                raw voltages on n channels. If False, it assumes either that
+                shots were acquired with the classifier detector (i.e. shots
+                are the probabilities of being in each state of the classifier)
+                or that they are raw voltages. Note that when preselection
+                the function checks for "classified_ro" and if it is false,
+                 (i.e. the input are raw voltages and not probas) then it uses
+                  the classifier on the preselection readouts regardless of the
+                  "predict_proba" flag (preselection requires classif of ground state).
```
As I understand, this flag tells the function, whether to do the classification or not. Wouldn't a name like `classify` or `do_classification` be more clear, since this is the term we use elsewhere in the code? Maybe there is some subtlety that I'm missing, which could then be more explicitly explained in the docstring.

nathlacroix replied:
Here predict_proba is more specific, it relates to outputing the probability of being in each of the n possible states for each single-shot measurement. I agree that this is, in fact, not what we want most of the  time (rather, we want to classify and then find the average probability). This relates to the "todo" line 1368. I think that went we implement that, we could change the name to "classify", which will then be appropriate.

antsr commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -1106,6 +1125,283 @@ def get_latex_prob_label(prob_label):
+            n_seqs = self.sp.length(1)  # corresponds to number of soft sweep points
+        else:
+            n_seqs = 1
+        # does not count preselection readout
+        n_readouts = list(shots_per_qb.values())[0].shape[0] // (n_shots * n_seqs)
```
What does it mean "does not count preselection readout"? From the code it does seem to include the presel. readout.

antsr replied:
Ok, I figured it out in the end and it looks to be working ok, but rather confusing to me.

Two suggestions to improve readability:
* The name of the flag `raw` in `_get_single_shots_per_qb`, could be `before_filtering` for example. Raw is not very specific.
* the variable `n_readouts` infers to number of readouts in a **sequence** after data filtering, not in a segment as I first thought.

nathlacroix replied:
I kept the 'raw' for consistency with "get_measured_data_per_qbs_raw".  (that is also where the data is taken from).  I added a note for the n_reaouds.

antsr commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -1106,6 +1125,283 @@ def get_latex_prob_label(prob_label):
+        # prepare preselection mask
+        if preselection:
+            # get preselection readouts
+            preselection_ro_mask = np.tile([True]*n_seqs + [False]*n_seqs,
+                                           n_shots*n_readouts )
```
This line seems to not support multiple readouts per segment.

chellings commented:
> @chellings you said you fixed something in the DynamicPhaseAnalysis but I didn't see any changes in that class_ in this _PR.

It was in the parent class (MultiCZgate_Calib_Analysis)
https://github.com/QudevETH/PycQED_py3/pull/262/commits/be8ba81a042e02fe2cfe08868b3fea4c3c31a3b0

stephlazar commented:
This analysis fails on Proj/S17 for a Rabi recorded with classified+thresholded single shots, no preselection. I had 33 sweep points but the analysis computes only 8 data points. This looks like a problem in how the single shots are averaged together in the analysis so maybe @nathlacroix can have a look.

nathlacroix committed:
Improve aesthetics in tda
- Improve semantic readability for data_filter
- Remove dupplicate row

nathlacroix committed:
Move code that finds the preselection qubits to parent class MultiCZgate_Calib_Analysis

nathlacroix committed:
Fix aesthetics and improve doc strings in tda

nathlacroix committed:
Remove dependency on number of acquisitions per segment when selecting the preselection readouts

Instead of assuming that there is one preselection readout and one readout per segment, it now assumes
one preselection readout and then computes the number of remaining readouts from the total number of readouts
before the data filtering.

nathlacroix committed:
Aesthetics and doc string improvements in tda

nathlacroix commented:
I should have answered/ improved all comments by now. @stephlazar could you try again your analysis and ensure that the appropriate kwargs are passed to the analysis for the measurement you speak of (e.g. data_type="singleshot", "classified_ro"=...., etc.). If it still doesn't work, we could have a look together

antsr commented:
My questions have been addressed and its ok to merge from my side now.

stephlazar commented:
@nathlacroix I can still not get this to work for the measurement I took. Let's discuss if I'm doing something wrong or there is a bug.

chellings committed:
Merge branch 'qudev_master' into enh/preselection_for_tda

nathlacroix commented:
I have checked your timestamp, @stephlazar . It seems like something went wrong in your measurement: you seem to be missing shots, which can be seen quite easily by looking at how many single-shot you acquired: 33792/33 sweep points = 1024 per sweep point. Yet in your qb4.acq_shots, you have 4096.  The detector metadata seems to be missing, so I cannot check further why you acquired only 25% of the data. I would conclude however, that the problem is not on the analysis side. Nonetheless, you can pass "n_shots=1024" in the options_dict, which seem to then yield a Rabi, see below
![Rabi_qb4_20210902_133959](https://user-images.githubusercontent.com/36669429/139603831-a8ee5a5d-d637-4bc3-bd43-4de2452a2063.png).
@stephlazar @chellings Let me know if you are ok to merge the pull request that stood open for longest in the history of pycqed (thanks for the review process, btw)

stephlazar commented:
@nathlacroix thanks for looking into it. Okay, strange that it didn't acquire all the shots ...
But from the analysis side it seems to work so I'm okay with merging this PR.
